lib/interceptor/cache.js

```diff
@@ -0,0 +1,138 @@
+'use strict'
+
+const CacheHandler = require('../handler/cache-handler.js')
+const LruCacheStore = require('../cache/lru-cache-store.js')
+
+/**
+ * Gives the downstream handler the request's cached response or dispatches
+ *  it if it isn't cached
+ * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions} globalOpts
+ * @param {*} dispatch TODO type
+ * @param {import('../../types/dispatcher.d.ts').default.RequestOptions} opts
+ * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+ * @param {import('../../types/cache-interceptor.d.ts').default.CacheStoreValue | undefined} value
+ */
+function handleCachedResult (
+  globalOpts,
+  dispatch,
+  opts,
+  handler,
+  value
+) {
+  const ac = new AbortController()
+  const signal = ac.signal
+
+  // Check value here as well since it still can be undefined if the store
+  //  returned a promise
+  if (!value) {
+    // Request isn't cached, let's continue dispatching it
+    dispatch(opts, new CacheHandler(globalOpts, opts, handler))
+    return
+  }
+
+  // TODO: check if the value is expired or not. If it is, then we need to send
+  //  a request to the origin revalidating it to make sure we can serve it
+  //  stale, and then also mark it as still-fresh somehow
+
+  // Request is cached, let's return it
+  try {
+    const {
+      statusCode,
+      statusMessage,
+      rawHeaders,
+      rawTrailers,
+      body,
+      cachedAt,
+      expiresAt
+    } = value
+
+    handler.onConnect(ac.abort)
+    signal.throwIfAborted()
+
+    // https://www.rfc-editor.org/rfc/rfc9111.html#name-age
+    const age = expiresAt - cachedAt / 1000
+    rawHeaders.push(Buffer.from(`age: ${age}`))
+
+    handler.onHeaders(statusCode, rawHeaders, () => {}, statusMessage)
+    signal.throwIfAborted()
+
+    if (opts.method === 'HEAD') {
+      handler.onComplete([])
+    } else {
+      for (const chunk of body) {
+        let ret = false
+        while (ret === false) {
+          ret = handler.onData(chunk)
+          signal.throwIfAborted()
+        }
+      }
+
+      handler.onComplete(rawTrailers)
+    }
+  } catch (err) {
+    handler.onError(err)
+  }
+}
+
+/**
+ * @param {import('../../types/cache-interceptor.d.ts').default.CacheOptions | undefined} globalOpts
+ * @returns {import('../../types/dispatcher.d.ts').default.DispatcherComposeInterceptor}
+ */
+module.exports = globalOpts => {
+  if (!globalOpts) {
+    globalOpts = {}
+  }
+
+  if (!globalOpts.store) {
+    globalOpts.store = new LruCacheStore()
+  }
+
+  if (!globalOpts.methods) {
+    globalOpts.methods = ['GET']
+  }
+
+  if (!globalOpts.statusCodes) {
+    globalOpts.statusCodes = [200]
+  }
+
+  return dispatch => {
+    return (opts, handler) => {
+      if (!globalOpts.methods.includes(opts.method)) {
+        // Not a method we want to cache, skip
+        return dispatch(opts, handler)
+      }
+
+      // Dump body
+      opts.body?.on('error', () => {}).resume()
```

Comment:
We should dump it, only if there's something cached isn't it?

Here we don't even now if cached and we already dumped the request's body